File: backend/app/models/uncertainty.py
"""Uncertainty representation for concept engineering — Phase 18."""

from __future__ import annotations

from enum import Enum

from pydantic import BaseModel, Field, model_validator

from app.models.concept import SourcedFloat, ValueSource


class Confidence(str, Enum):
    """How much weight the estimate can carry."""

    #: The range comes from a cited figure or from arithmetic on known
    #: values; the working value is unlikely to be far wrong.
    HIGH = "HIGH"
    # A defensible preliminary assumption for a familiar operation.
    MEDIUM = "MEDIUM"
    # A first-pass figure to make the model runnable at all.
    LOW = "LOW"


class EstimateMethod(str, Enum):
    """How the number was actually arrived at. Never decorative."""

    # Arithmetic on values the concept already holds (e.g.
    DERIVED = "DERIVED"
    # A bundled reference dataset, named in `basis`.
    REFERENCE_DATA = "REFERENCE_DATA"
    # A language model structured the engineer's description into a range.
    LANGUAGE_MODEL = "LANGUAGE_MODEL"
    #: Composed from Fabrivium's documented reference bands when the
    #: language model could not be reached (Phase 18B). The value is still
    #: an ENGINEERING_ESTIMATE — only the mechanism differs.
    LOCAL_HEURISTIC = "LOCAL_HEURISTIC"
    # The engineer typed the range themselves.
    ENGINEER = "ENGINEER"


class EstimatedRange(BaseModel):
    """A preliminary engineering assumption expressed as a range."""

    model_config = {"frozen": True}

    low: float = Field(..., gt=0, description="Optimistic end of the range")
    working_value: float = Field(..., gt=0, description="The value the simulation will use")
    high: float = Field(..., gt=0, description="Pessimistic end of the range")
    unit: str = Field(..., min_length=1, description="e.g. 's', 'units'")

    confidence: Confidence
    method: EstimateMethod
    #: What the number rests on, in the engineer's words — "6 fastening
    #: operations + handling allowance". An estimate without a basis is
    #: indistinguishable from a guess, so this is required.
    basis: str = Field(..., min_length=1)
    #: Only set when a language model was involved, so the provenance view
    #: can name it rather than saying "AI".
    model_name: str | None = None
    # G11 — how many times the operation was assumed to happen per unit when this range
    # was composed.
    operations_per_unit: int | None = None

    @model_validator(mode="after")
    def _ordered_and_contained(self) -> "EstimatedRange":
        if self.low > self.high:
            raise ValueError(f"Estimated range is inverted: low {self.low} > high {self.high}.")
        if not (self.low <= self.working_value <= self.high):
            # A working value outside its own range means the two disagree
            # about what is being claimed, and the KPIs would carry a number
            # the range does not admit.
            raise ValueError(
                f"Working value {self.working_value} lies outside the estimated "
                f"range {self.low}–{self.high}."
            )
        return self

    @property
    def spread(self) -> float:
        return self.high - self.low

    def resolve(self, *, detail: str | None = None) -> SourcedFloat:
        """The single scalar the deterministic pipeline will receive."""
        return SourcedFloat.of(
            self.working_value,
            ValueSource.ENGINEERING_ESTIMATE,
            detail or f"{self.low:g}–{self.high:g} {self.unit}, {self.confidence.value.lower()} confidence · {self.basis}",
        )

    def sweep_points(self) -> list[float]:
        """The values a sensitivity sweep should evaluate."""
        seen: list[float] = []
        for value in (self.low, self.working_value, self.high):
            if not any(abs(value - kept) < 1e-9 for kept in seen):
                seen.append(value)
        return sorted(seen)


class EstimateUnavailable(BaseModel):
    """Why no estimate could be produced."""

    model_config = {"frozen": True}

    reason: str = Field(..., min_length=1)
    #: True when a human could fix this by trying later or configuring a
    #: provider; false when no amount of retrying would help.
    retryable: bool = False


class ValueRevision(BaseModel):
    """One step in a value's life, kept so a change can be explained."""

    model_config = {"frozen": True}

    field: str = Field(..., min_length=1, description="e.g. 'cycle_time'")
    previous_value: float | None = None
    previous_source: ValueSource
    new_value: float
    new_source: ValueSource
    reason: str = Field(..., min_length=1)
    # Where the stronger evidence came from, when there is a document.
    evidence_url: str | None = None

    @model_validator(mode="after")
    def _must_be_a_real_change(self) -> "ValueRevision":
        if self.previous_value == self.new_value and self.previous_source is self.new_source:
            raise ValueError("A revision that changes neither the value nor its source is not a revision.")
        return self


# The order in which evidence supersedes evidence.
EVIDENCE_STRENGTH: dict[ValueSource, int] = {
    ValueSource.UNKNOWN: 0,
    # A stated planning convention, applying to no particular factory.
    ValueSource.CATALOG_DEFAULT: 1,
    # Preliminary assumptions. Good enough to simulate a concept with.
    ValueSource.ENGINEERING_ESTIMATE: 2,
    ValueSource.EXAMPLE_DATA: 2,
    # Fabrivium's own predictions — arithmetic, or a simulation run.
    ValueSource.CALCULATED: 3,
    ValueSource.SIMULATED: 3,
    # Facts about a specific real item, published by whoever sells it.
    ValueSource.MANUFACTURER: 4,
    ValueSource.EXTERNAL_DATA: 4,
    # A person's statement, decision or cited paperwork. See the tie note.
    ValueSource.CUSTOMER: 5,
    ValueSource.ENGINEER: 5,
    ValueSource.DOCUMENT: 5,
    # Observed on a real line. The only entry that is not a prediction.
    ValueSource.MEASURED: 6,
}

# Ranks that mean "a person or an instrument settled this".
DECIDED_BY_A_PERSON = 5


def is_upgrade(previous: ValueSource, new: ValueSource) -> bool:
    """Whether `new` is stronger evidence than `previous`."""
    return EVIDENCE_STRENGTH[new] > EVIDENCE_STRENGTH[previous]


class StationAssumptionProposal(BaseModel):
    """Preliminary simulation assumptions for ONE station — Phase 18B."""

    model_config = {"frozen": True}

    stage_id: str = Field(..., min_length=1)
    stage_name: str = Field(..., min_length=1)

    # Seconds per unit. None when neither route could justify a figure.
    cycle_time: EstimatedRange | None = None
    #: Units the station processes CONCURRENTLY — the simulator's
    #: `simpy.Resource` capacity, not batch size and not buffer space.
    capacity: EstimatedRange | None = None
    # Operators this station occupies from the factory pool while running.
    operators: EstimatedRange | None = None

    #: True when the language model could not be reached and the local
    #: heuristic produced these.
    fell_back: bool = False
    # Provider-side detail, for developers. Never the headline.
    provider_note: str | None = None

    @property
    def proposed_fields(self) -> list[str]:
        """Which parameters actually carry a proposal."""
        return [
            name
            for name, value in (
                ("cycle_time", self.cycle_time),
                ("capacity", self.capacity),
                ("operators", self.operators),
            )
            if value is not None
        ]

    @property
    def has_any(self) -> bool:
        return bool(self.proposed_fields)

File: frontend/src/api/uncertainty.ts
import { apiPost } from "./client";
import type { FactoryConceptDraft, ConceptResponse, SourcedNumber } from "./types";

/** Phase 18 — uncertainty-aware concept assistance. */

export type Confidence = "HIGH" | "MEDIUM" | "LOW";
export type EstimateMethod =
  | "DERIVED"
  | "REFERENCE_DATA"
  /** A language model structured the description into a range. */
  | "LANGUAGE_MODEL"
  /** Composed from Fabrivium's documented reference bands (Phase 18B). */
  | "LOCAL_HEURISTIC"
  | "ENGINEER";

export interface EstimatedRange {
  low: number;
  working_value: number;
  high: number;
  unit: string;
  confidence: Confidence;
  method: EstimateMethod;
  /** What the number rests on. An estimate without one is a guess. */
  basis: string;
  model_name: string | null;
  /** How many times the operation was assumed to happen per unit when this
   * range was composed — propagated from the reviewed process, typed by the
   * engineer, or read out of the description. Travels back with the
   * acceptance so the station keeps the assumption it was estimated under
   * (G11). Absent where the count played no part. */
  operations_per_unit?: number | null;
}

export interface EstimateUnavailable {
  reason: string;
  retryable: boolean;
}

/** How many of the concept's values came from where. */
export interface ProvenanceCounts {
  customer_facts: number;
  manufacturer_facts: number;
  engineering_estimates: number;
  example_data: number;
  derived: number;
  planning_defaults: number;
  unknown: number;
  engineer_decisions: number;
  documents: number;
  measured: number;
  external_data: number;
  simulated: number;
}

export interface ConceptReadiness {
  counts: ProvenanceCounts;
  simulation_ready: boolean;
  verdict: string;
  unknown_critical: number;
  unknown_commercial: number;
  missing: string[];
  /** Definitional arithmetic — an upper bound on the average cycle, never
   * an achievable per-station time. */
  takt_seconds: SourcedNumber;
}

/** What Fabrivium still needs before it can estimate. */
export interface NeedsInformation {
  reason: string;
  questions: string[];
}

/** The description and the selected automation level disagree. */
export interface EstimateContradiction {
  message: string;
  described_as: string;
  selected_as: string;
}

export interface EstimateResult {
  /** Exactly one of these three is set. */
  estimate: EstimatedRange | null;
  /** The whole station, when one could be proposed. */
  proposal: StationAssumptionProposal | null;
  needs_information: NeedsInformation | null;
  contradiction: EstimateContradiction | null;
  /** True when the language model could not be reached and the local
   * heuristic produced the range instead. */
  fell_back: boolean;
  /** Provider-side detail for developers. Never the headline: the demo
   * does not show quota internals. */
  provider_note: string | null;
  takt_seconds: SourcedNumber;
}

export interface SweepPoint {
  value: number;
  unit: string;
  completed_units: number;
  target_units: number;
  meets_target: boolean;
  bottleneck_machine_id: string;
}

export interface SensitivityResult {
  stage_id: string;
  stage_name: string;
  parameter: string;
  unit: string;
  points: SweepPoint[];
  /** Real runs of the deterministic simulator — not samples. */
  simulations_run: number;
  monotonic: boolean;
  summary: string;
}

export interface ThresholdResult {
  stage_id: string;
  stage_name: string;
  parameter: string;
  unit: string;
  /** Null whenever no honest single number exists; `statement` says why. */
  threshold: number | null;
  target_units: number;
  simulations_run: number;
  monotonic: boolean;
  statement: string;
  requirement_value: SourcedNumber;
}

/** What the concept is made of, by provenance. No completeness score. */
export function conceptReadiness(draft: FactoryConceptDraft): Promise<ConceptReadiness> {
  return apiPost<ConceptReadiness>("/concept/readiness", { draft });
}

/** Ask the assistant for a preliminary cycle-time range. */
export function estimateCycleTime(
  draft: FactoryConceptDraft,
  stageId: string,
  input: {
    description: string;
    automation_level?: string;
    operations_per_unit?: number | null;
    part_information?: string | null;
    other_constraints?: string | null;
    /** AUTO (default) | LLM_ONLY | LOCAL_ONLY. */
    mode?: string | null;
  },
): Promise<EstimateResult> {
  return apiPost<EstimateResult>("/concept/estimate", {
    draft,
    stage_id: stageId,
    description: input.description,
    automation_level: input.automation_level ?? "UNKNOWN",
    operations_per_unit: input.operations_per_unit ?? null,
    part_information: input.part_information ?? null,
    other_constraints: input.other_constraints ?? null,
    mode: input.mode ?? null,
  });
}

/** Write an estimate's working value onto a stage, tagged as an estimate. */
export function applyEstimate(
  draft: FactoryConceptDraft,
  stageId: string,
  range: Omit<EstimatedRange, "unit"> & { unit?: string },
  /** Required before an estimate may replace a value that came from a
   * person, a document, a measurement or a manufacturer. Without it the
   * backend answers 409 and names what would have been lost — see
   * `ProtectedValueConflict`. */
  replaceExisting = false,
): Promise<ConceptResponse> {
  return apiPost<ConceptResponse>("/concept/apply-estimate", {
    draft,
    stage_id: stageId,
    low: range.low,
    working_value: range.working_value,
    high: range.high,
    basis: range.basis,
    confidence: range.confidence,
    method: range.method,
    model_name: range.model_name,
    replace_existing: replaceExisting,
  });
}

/** One value an estimate would have overwritten. */
export interface ProtectedValue {
  field: string;
  label: string;
  value: number | null;
  source: string;
  detail: string | null;
}

/** The backend's refusal to overwrite something stronger than an estimate. */
export interface ProtectedValueConflict {
  conflict: "PROTECTED_VALUE";
  /** Readable on its own, so the generic error path degrades gracefully if
   * a caller does not handle the conflict. */
  message: string;
  protected: ProtectedValue[];
}

/** Reads a 409 from either apply path, or null for any other failure. */
export function protectedValueConflict(error: unknown): ProtectedValueConflict | null {
  const candidate = error as { status?: number; detail?: unknown };
  if (candidate?.status !== 409) return null;
  const detail = candidate.detail as ProtectedValueConflict | undefined;
  if (detail?.conflict !== "PROTECTED_VALUE") return null;
  return detail;
}

/** Run the deterministic simulator once per value. A bounded sweep — not
 * sampling, and not Monte Carlo. */
export function runSensitivity(
  draft: FactoryConceptDraft,
  stageId: string,
  values?: number[],
): Promise<SensitivityResult> {
  return apiPost<SensitivityResult>("/concept/sensitivity", {
    draft,
    stage_id: stageId,
    values: values ?? null,
  });
}

/** What this parameter must achieve for the concept to meet its target. */
export function deriveThreshold(
  draft: FactoryConceptDraft,
  stageId: string,
  bounds?: { fastest?: number; slowest?: number },
): Promise<ThresholdResult> {
  return apiPost<ThresholdResult>("/concept/threshold", {
    draft,
    stage_id: stageId,
    fastest: bounds?.fastest ?? 5,
    slowest: bounds?.slowest ?? 120,
  });
}

/** Phase 18B — one station, every simulation parameter. */
export interface StationAssumptionProposal {
  stage_id: string;
  stage_name: string;
  cycle_time: EstimatedRange | null;
  /** Units in process AT THE SAME TIME — the simulator's server capacity,
   * not batch size and not buffer space. */
  capacity: EstimatedRange | null;
  /** People this station occupies while it runs. Never the size of the factory's workforce pool. */
  operators: EstimatedRange | null;
  fell_back: boolean;
  provider_note: string | null;
}

export interface AcceptAssumptionsResult {
  draft: FactoryConceptDraft;
  validation: unknown;
  /** What was actually written, so the UI reports the truth rather than what it asked for. */
  applied: string[];
}

/** Write the accepted parameters into the concept. */
export function acceptStationAssumptions(
  draft: FactoryConceptDraft,
  proposal: StationAssumptionProposal,
  acceptedFields: string[],
  /** As `applyEstimate`: accepting an assumption over a value a person is
   * accountable for takes a second, explicit act. */
  replaceExisting = false,
): Promise<AcceptAssumptionsResult> {
  return apiPost<AcceptAssumptionsResult>("/concept/accept-assumptions", {
    draft,
    proposal,
    accepted_fields: acceptedFields,
    replace_existing: replaceExisting,
  });
}
